# TWITTER ANALYSIS

**Report Content**  
1.Introduction Section: ⁃ Discussion about the analysis.  
2.Data Section: ⁃ Description of the data that will be used to solve the problem and the sources.  
3.Methodology section ⁃ The Methods are used extract the Information.  
5.Results section ⁃ Discussion of the results.  
6.Discussion section ⁃ Elaboration and discussion.  
7.Conclusion section

**1) Introduction** In the time, when everyone is going through the situation because of ongoing pandemic, people are also expressing their emotions on Twitter, its a social platform where users share informations like news, their experiences, they ask for help, they provide the help, they discuss about problems they are facing, and also get the solutions there. It has become a voice of the mass. So to study the information related to covid-19 from twitter an analysis has been done in the project.

**2) Data** Data set was provided for the analysis in the form of CSV file.

In [2]:
!pip install xlrd

In [3]:
import pandas as pd
print ("done")

done


In [5]:
import csv

In [6]:
df = pd.read_csv('https://kashifa-dataset.s3-ap-southeast-1.amazonaws.com/covid19_tweets.csv')

In [7]:
df.head(3)


user_name     user_location  \
0           ᏉᎥ☻լꂅϮ        astroworld   
1    Tom Basile 🇺🇸      New York, NY   
2  Time4fisticuffs  Pewee Valley, KY   

                                    user_description         user_created  \
0  wednesday addams as a disney princess keepin i...  2017-05-26 05:46:42   
1  Husband, Father, Columnist & Commentator. Auth...  2009-04-16 20:06:23   
2  #Christian #Catholic #Conservative #Reagan #Re...  2009-02-28 18:57:41   

   user_followers  user_friends  user_favourites  user_verified  \
0             624           950            18775          False   
1            2253          1677               24           True   
2            9275          9525             7254          False   

                  date                                               text  \
0  2020-07-25 12:27:21  If I smelled the scent of hand sanitizers toda...   
1  2020-07-25 12:27:17  Hey @Yankees @YankeesPR and @MLB - wouldn't it...   
2  2020-07-25 12:27:14  @diane3443 @wdunlap @realDonaldTrump Trump nev...   

      hashtags               source  is_retweet  
0          NaN   Twitter for iPhone       False  
1          NaN  Twitter for Android       False  
2  ['COVID19']  Twitter for Android       False

In [8]:
total_tweets= len(df[['user_name']])
print("total tweet:", total_tweets)

total tweet: 179108


In the data set we have 179108 tweets information to analyse. It contains the informations about user name, location, description, date when the account was created, followers, friends, is it verified, date of tweet, text the tweet contains, hashtags used for tweets, source of the tweet, and if its reweeted. Based on the information analysis was done. 

**3) Methodology:**  To extract the information from the data various functions have been used and presented quite a few insights from the information. 

#### INSIGHTS FOR DATA:

**i) USER LOCATION:** for the first insight, user location was selected, It would further provide the locations for major tweets and users as well.

**a)** it was to count the total locations from where people tweeted about the covid-19

In [9]:
x= df[['user_location']].drop_duplicates()
Total_locations=len(x['user_location'])
print("total locations:",Total_locations)
x.head()

total locations: 26921


user_location
0            astroworld
1          New York, NY
2      Pewee Valley, KY
3  Stuck in the Middle 
4     Jammu and Kashmir

from total 26921 locations tweets were tweeted.

**b)** it was to count for hightest number of tweets from the locations

In [10]:
df['user_location'].value_counts()

India                             3741
United States                     2455
New Delhi, India                  1721
Mumbai, India                     1401
Washington, DC                    1354
London, England                   1269
United Kingdom                    1206
London                            1103
New York, NY                      1082
Australia                          964
Worldwide                          879
Canada                             835
New Delhi                          822
USA                                817
UK                                 803
WORLDWIDE                          777
Los Angeles, CA                    776
South Africa                       766
Lagos, Nigeria                     731
California, USA                    709
Global                             688
Nairobi, Kenya                     618
Nigeria                            579
Atlanta, GA                        578
Chicago, IL                        561
Switzerland              

There are 3741 tweets from 26920 total locations used **India** as their location. There are few more locations based in India were used to tweet.

**C)** To find the number of user from every location

In [11]:
df_a=df[["user_name","user_location"]]

In [12]:
df_b=df_a.drop_duplicates(subset="user_name",keep="first")
df_b.head(2)

user_name user_location
0         ᏉᎥ☻լꂅϮ    astroworld
1  Tom Basile 🇺🇸  New York, NY

In [13]:
df_c = df_b["user_location"].value_counts().rename_axis('location').reset_index(name='users')
df_c.head(3)

location  users
0    United States   1277
1            India   1044
2  London, England    776

united states has 1277 users who tweeted about covid, which is the highest among all other location

**2) VERIFIED USERS**: There are well known people and others hear them and follow them, twitter usually provide them a verified account, both verified and non verified accounts are active,, so to check the part of tweets from verified accounts on covid.

In [14]:
dfverified= df[df['user_verified']==True]
dfv=dfverified.drop_duplicates(subset='user_name', keep= 'first')
verified_accounts= len(dfv['user_name'])
print("verified accounts:", verified_accounts)
dfv.head(3)

verified accounts: 6224


user_name  user_location  \
1        Tom Basile 🇺🇸   New York, NY   
21   APO Group English  #AFRICA #MENA   
25  Blood Donors India  Mumbai, India   

                                     user_description         user_created  \
1   Husband, Father, Columnist & Commentator. Auth...  2009-04-16 20:06:23   
21  Latest #Africa & #MENA related #News releases ...  2011-02-22 09:09:45   
25  Focused on matching blood donors with those in...  2008-12-23 07:55:39   

    user_followers  user_friends  user_favourites  user_verified  \
1             2253          1677               24           True   
21           10661             6             2037           True   
25         1215920          2047            19359           True   

                   date                                               text  \
1   2020-07-25 12:27:17  Hey @Yankees @YankeesPR and @MLB - wouldn't it...   
21  2020-07-25 12:26:31  Coronavirus - South Africa: COVID-19 update fo...   
25  2020-07-25 12:26:26  #Kolar\nNeed #Blood Type :  B-positive\nAt : J...   

                         hashtags               source  is_retweet  
1                             NaN  Twitter for Android       False  
21                            NaN      Africa Newsroom       False  
25  ['Kolar', 'Blood', 'COVID19']   Blood Donors India       False

There are 6,224 accounts are verified.

**b)** Verified accounts in different location

In [15]:
df_v=dfv[["user_name","user_location"]]
df_v.head(3)

user_name  user_location
1        Tom Basile 🇺🇸   New York, NY
21   APO Group English  #AFRICA #MENA
25  Blood Donors India  Mumbai, India

In [16]:
df_vl=df_v["user_location"].value_counts().rename_axis('location').reset_index(name='verified_users')
df_vl.head(5)

location  verified_users
0    Washington, DC             186
1             India             106
2      New York, NY              85
3  New Delhi, India              80
4            London              76

out of 6224, 186 verified users are in washinton, DC, where 106 in India and 80 gives Delhi India as there location. people are using different location type like country and cities, so on the basis of above data it seems that the most verified users are tweeting about covid are from washington, so the washington contains the highest number of verified user if comparing with other "CITIES". 

**3) HASHTAGS:** Various hashtags have been used in the data. Here, To collect the information about hashtags.

**a)** To know how many people used hashtags and how many people tweeted without hashtags.

In [18]:
dfH=df.dropna()
tweets_with_hashtag=len(dfH[['user_name']])
tweets_without_hashtag=total_tweets-tweets_with_hashtag
print("tweets with hashtag:",tweets_with_hashtag )
print("tweets without hashtag:",tweets_without_hashtag )

tweets with hashtag: 99138
tweets without hashtag: 79970


There are 99138 tweets were tweeted with hashtags and 79970 are without hashtags.

**b)** To know which hashtag has been used most frequently 

In [19]:
dfH['hashtags'].value_counts()

['COVID19']                                                                                30054
['Covid19']                                                                                 3838
['covid19']                                                                                 2368
['coronavirus']                                                                              442
['COVID19', 'coronavirus']                                                                   407
['coronavirus', 'COVID19']                                                                   388
['Coronavirus', 'COVID19']                                                                   267
['CoronaVirusUpdates', 'COVID19']                                                            245
['Coronavirus']                                                                              230
['India', 'COVID19']                                                                         147
['IndiaFightsCorona', 'COVID19

There are so many hashtags were used related to covid, Out of 41709 Hashtags COVID19 alone was used 30054 times. 

**C)** Hashtags used by verified account

In [20]:
dfv1=dfv.dropna()
verified_accounts_used_hashtags=len(dfv1['user_name'])
print("number of verified account use hashtags:", verified_accounts_used_hashtags)
dfv1['hashtags'].value_counts()

number of verified account use hashtags: 4098


['COVID19']                                                              1781
['Covid19']                                                               187
['covid19']                                                                94
['coronavirus', 'COVID19']                                                 18
['coronavirus']                                                            18
['COVID19', 'coronavirus']                                                 16
['COVID19', 'pandemic']                                                     8
['BREAKING', 'COVID19']                                                     8
['PASenate', 'COVID19']                                                     7
['ICYMI']                                                                   7
['COVID19', 'COVID19']                                                      5
['COVID19', 'asthma']                                                       5
['ICYMI', 'COVID19']                                            

So, out of 6224 verified accounts 4098 verified accounts used hashtags and COVID19 was the most used hashtag with 1781 tweets, by the information saying that people use the hashtags used by verified users and verified users used the most used hashtags to tweet is quite justifying with the situation.

**4) USER NAME:** to know the most frequent users

**a)** Total number of users tweeted.

In [21]:
dfU=df[['user_name']].drop_duplicates()
Total_number_of_users=len(dfU)
print("Total number of users:",Total_number_of_users)

Total number of users: 92276


There are 92276 users who tweeted about covid with and without hashtags.

**b)** Frequent user who tweeted the most.

In [22]:
dfA=df["user_name"].value_counts()
dfA.head(5)

GlobalPandemic.NET     679
Coronavirus Updates    625
covidnews.ch           402
Open Letters           390
Blood Donors India     282
Name: user_name, dtype: int64

So, "GlobalPandemic.NET" has tweeted 679 times, which is the highest number of tweets by any user, after that "Coronavirus Updates" and "covidnews.ch"  with 625 and 402 tweets.

**C)** Most frequent verified user 

In [23]:
dfc=dfverified['user_name'].value_counts()
dfc.head(3)

Open Letters          390
Blood Donors India    282
Hindustan Times       280
Name: user_name, dtype: int64

So, "Open letters" tweeted the most with 390 tweets among all verified accounts

**5) Date of Tweets:** To check the dates where tweets had been done the most.

In [24]:
dfD=df[["user_name","date"]]       #dataframe with two columns only user's name and the date of tweet

In [25]:
print(dfD["date"].dtypes)         #type of date column

object


In [26]:
df_users = pd.DataFrame()
df_users['date'] = pd.to_datetime(dfD['date']).dt.date             #changed the type in date time and again in object to remove the time from the date
df_users['user_name'] = dfD["user_name"]

In [27]:
print(df_users["date"].dtypes)

object


In [28]:
df_users[['user_name','date']].head(2)                      
#now how this is looking like without time, adding time will creat the dataframe much clustered so to make it more feasiable removed the time

user_name        date
0         ᏉᎥ☻լꂅϮ  2020-07-25
1  Tom Basile 🇺🇸  2020-07-25

In [29]:
df_tweets = df_users["date"].value_counts().rename_axis('dates').reset_index(name='tweets')      #counting the tweets on every date
df_tweets.head(3)

dates  tweets
0  2020-07-25   16881
1  2020-08-22   11555
2  2020-08-30    8375

In [30]:
df_dashbord=df_tweets.sort_values(by="dates")
df_dashbord.head(3)                                          #sorted the data by date from 24/7 onwards

dates  tweets
25  2020-07-24     295
0   2020-07-25   16881
12  2020-07-26    7500

In [31]:
import pandas as pd
from bokeh.plotting import figure, output_file, show,output_notebook
output_notebook()

Loading BokehJS ...

In [32]:
def make_dashboard(DATE, TWEETS, title, file_name):
    output_file(file_name)
    p = figure(title=title, x_axis_label='Date', y_axis_label='Tweets')
    p.line(DATE.squeeze(), TWEETS.squeeze(), color="firebrick", line_width=4, legend="tweets")
    show(p)                                                                          #making dashboard

In [33]:
DATE=df_dashbord[["dates"]]                     

In [34]:
TWEETS=df_dashbord[["tweets"]]

In [35]:
title="Tweets_Dashboard"

In [36]:
file_name="tweets.html"

In [37]:
make_dashboard(DATE=DATE, TWEETS=TWEETS, title=title, file_name=file_name)

Most of the tweets were tweeted in the beginning, later it become constant and few times went lesser and sometime came a bit more, it is predictable from the trend that the sentiments on twitter were much higher in july the way people were facing the situation, but later they became used to of it or the situations are better for them so tweeted much lesser and the number of tweets everyday became constant.

**6) SOURCES:** Sorces were used to tweet and to use hashtags

In [38]:
source=df["source"].value_counts()
source.head(5)

Twitter Web App        56891
Twitter for Android    40179
Twitter for iPhone     35472
TweetDeck               8543
Hootsuite Inc.          7321
Name: source, dtype: int64

The busiest source,Covid-19 was discused most is Twitter web app, 56891 users are using Twitter web app.

**7) Trend of followers** for the insight, it need to be understand how people are following other tweeter account

In [39]:
df.head(2)

user_name user_location  \
0         ᏉᎥ☻լꂅϮ    astroworld   
1  Tom Basile 🇺🇸  New York, NY   

                                    user_description         user_created  \
0  wednesday addams as a disney princess keepin i...  2017-05-26 05:46:42   
1  Husband, Father, Columnist & Commentator. Auth...  2009-04-16 20:06:23   

   user_followers  user_friends  user_favourites  user_verified  \
0             624           950            18775          False   
1            2253          1677               24           True   

                  date                                               text  \
0  2020-07-25 12:27:21  If I smelled the scent of hand sanitizers toda...   
1  2020-07-25 12:27:17  Hey @Yankees @YankeesPR and @MLB - wouldn't it...   

  hashtags               source  is_retweet  
0      NaN   Twitter for iPhone       False  
1      NaN  Twitter for Android       False

In [40]:
df_follower=df[["user_name","user_followers","user_verified"]]

**a)** to find the most following twitter account

In [41]:
df_follower.sort_values(by="user_followers", ascending=False)


user_name  user_followers  user_verified
104946                        CNN        49442559           True
174909        National Geographic        24359165           True
6959                         CGTN        13892841           True
13450                        CGTN        13892839           True
16194                        CGTN        13892837           True
235                          CGTN        13892795           True
2837                         CGTN        13892793           True
5344                         CGTN        13892792           True
23721                        CGTN        13892212           True
20378                        CGTN        13892212           True
20483                        CGTN        13892212           True
24243                        CGTN        13892212           True
17823                        CGTN        13892211           True
36181                        CGTN        13890039           True
38250                        CGTN        13890038           True
38948                        CGTN        13890036           True
53024                        CGTN        13885511           True
67481                        CGTN        13882345           True
69728                        CGTN        13882344           True
73686                        CGTN        13882338           True
77882                        CGTN        13880255           True
89642                        CGTN        13878882           True
94963                        CGTN        13878480           True
90559                        CGTN        13878480           True
93043                        CGTN        13878479           True
92455                        CGTN        13878479           True
94233                        CGTN        13878479           True
101910                       CGTN        13877202           True
105491                       CGTN        13876219           True
105918                       CGTN        13876219           True
...                           ...             ...            ...
159390                AsherSunset               0          False
53690                         Cad               0          False
176055            TurdShirt Sales               0          False
100287             David Ferguson               0          False
159336    helalbatstuta@gmail.com               0          False
8595                   ArunKantha               0          False
159293            Lancashire Lass               0          False
100144     Perpetual Motion Drone               0          False
54006                  kimberly s               0          False
159485            आत्मनिर्भर लेबर               0          False
159516                  neha garg               0          False
53241            david derbyshire               0          False
52872          RidiculousRhetoric               0          False
52527                     NEastly               0          False
52548                Gerardo Rios               0          False
134127                    Zelazo🤡               0          False
8978                     Droupons               0          False
101010                       Kyra               0          False
143067         Juhi Bandyopadhyay               0          False
159661                  neha garg               0          False
159651           Muzammil Piyarji               0          False
53204                       Peter               0          False
134092               Polly Rhymer               0          False
159621              rakesh pareek               0          False
8865                      MagaDad               0          False
127964  Cygnus Professionals Inc.               0          False
175996       Anna Lisa Valderrama               0          False
53096             Covid π = Death               0          False
159549       Mohammed Zaid khilji               0          False
9808             blinkbloomdesign               0          

CNN has the largest number of followers.

**b)** To know the relation between person held followers and tweeting about

In [42]:
df_twt=df_follower["user_name"].value_counts().rename_axis('user_name').reset_index(name='tweets')
df_twt.head(3)                                          #finding number of tweets

user_name  tweets
0   GlobalPandemic.NET     679
1  Coronavirus Updates     625
2         covidnews.ch     402

In [43]:
df_temp=df[["user_name","user_followers"]].drop_duplicates("user_name")
df_f=df_temp.sort_values(by="user_followers", ascending=False)
df_f.head(3)                     #finding followers of users

user_name  user_followers
104946                  CNN        49442559
174909  National Geographic        24359165
235                    CGTN        13892795

In [44]:
def make_dashboard(followers, tweets, title, file_name):
    output_file(file_name)
    p = figure(title=title, x_axis_label='Follower', y_axis_label='Tweets')
    p.line(followers.squeeze(), tweets.squeeze(), color="firebrick", line_width=4, legend="tweets")
    show(p)

In [45]:
followers=df_f[["user_followers"]]

In [46]:
tweets=df_twt[["tweets"]]

In [47]:
title="Tweet_follower"

In [48]:
file_name="followers.html"

In [49]:
make_dashboard(followers=followers, tweets=tweets, title=title, file_name=file_name)

The above graph is showing the people have more followers have tweeted more tweets.

**4)Results:**
1) In the data 26921 location were mentioned for tweeting about covid and most of the tweets were from India. But most of the users were from US. Us has a large number of user expressing their sentiments.  
2) The number of verified users is 6224 out of which 186, is  highest number than any other city are from washington DC.  
3) people usually use hashtags with the tweet, according to data 99138 tweets were with hashtags, but here it is a good number of tweets didnt use the hashtag and thats 79970, hashtag COVID19 alone was used mostly with 30054 and it was use by 1781 verified user.  
4) There are 92276 users who tweeted in the dataset, "GlobalPandemic.NET" was most active and tweeted 679 times and among all verified users open letter tweeted 390 tweets which was highest.  
5) most of the tweets were from the beginning of the data i.e. in july end, then became constant in the number with the days.  
6) Mostly people used twitter web app to tweet as a source,  it was used by 56891 tweets.  
7) among all users, CNN has the largest number of followers that's 49442559, from trends the people who tweeted the most have more number of followers.

**6) conclusion**  
1) US has more number of users tweeting about covid, while indians have more tweets. So, indians are expressing much sentiments over Covid it can be concluded that Indians are much affected by the situation.  
2) Verified users uses hashtags more frequently for showing their expression, verified users from washingtong, DC are more involved in expressing the sentiments than any other city.  
3) Users used COVID19 hashtag with the highest frequency, even verified users used the hashtag, there is a connection says if verified users uses the hashtag it becomes more popular.  
4) In all users Globalpandemic.Net was tweeted the most where a verified account named open letter has highest number of tweets in among all verified users, Globalpandemic.Net is not verified account but tweeted more than any verified account.  
5) It is predictable from the trend that the sentiments were much higher in july the way people were facing the situation, but later they became used to of it or the situations are better for them so tweeted much lesser and the number of tweets everyday became constant.  
6) people used tweeter web app to tweet, there are news portal's official accounts tweet about, hence the number is high of using Twitter web app and most of the companies giving work from home as well so users at home while working on there computers also using the twitter web app.  
7) trends says that users with more followers tweeted more than other users while comparing.

**7) discussion**
Here presented a few number of insights but the data has really large number of information, more predictions could be possible to extract from the data. here are few examples:  
1) data has used blood donor with covid by tracking the locations of users asking donor it is possible to predict how many petients are recovered and how many still going through and also which location is using plasma therapy to treat people the most, and how people are reacting towards, like in which number they are accepting this treatment.  
2) By the information of retweets it is predictable which tweets has retweeted most, and the role of hashtag in retweeting is the highest no of hashtag result of retweets of the hashtag.  
3) by user's account created date it is possible to provide the information that how many accounts have been made during the pandemic and how many before that, if these recent account tweeting more than old accounts, like there was a account name global pandemic tweeted the most by the name seems a new account has tweeted the most but it is possible that it changed its name so to confirm that it is needed to work with created date.  

there many other informations are possible from the data.

-----------Thank you-------------